Readme.md
# nwn_2da_tlkify

A tool to assemble 2das/tlks from a collection of json files.


##Configure operation in one of two ways:

1) Command line (has preference)

```
  -i TLK                      Input TLK file, Config key: tlk
  -s START                    Starting StrRef, Config key: start [default: highest]
  -o OUTTLK                   Output TLK file file, Config key: outtlk
  -d INDIR                    2da input directory, Config key: twoda
  -a OUTDIR                   2da output directory Config key: outtwoda
  -j JSONDIR                  Json Input directory, Config key: tlkjson
  -c CONFIG                   Configuration file [default: 2tconfig.ini]
  --auto-lower SEL            Turns on, off, or on for a specific 2da the automatically setting
                              the Lower field in classes or ConverNameLower field in racialtypes
                              Possible selections: all, off, class, and race.
                              Config key: autolower [default: all]
  --auto-plural PLU           Turns on, off, or on for a specific 2da the automatically setting
                              the Plural field in classes or NamePlural field in racialtypes
                              Possible selections: all, off, class, and race.
                              Config key: autoplural [default: all]
  --auto-adjective            Turns on auto setting of the ConverName from Name in racialtypes.
                              Config key: autoadjective true/false
  --auto-iprp_spells ISP      Sets the iprp_spells row to start autofill from the spells.2da.
                              Config key: autoiprp_spells [default: -1]
  --auto-spells SID           Sets the spells row to start autofill from feat.2da for Name and SpellDesc columns.
                              Config key: autospells [default: -1]
  --auto-iprp_feats FID       Sets the iprp_feats row to start autofill from feat.2da for the Name column.
                              Config key: autoiprp_feats [default: -1]
  --reserv-spells RID         Reserves space for spells.2da Name/SpellDesc columns so that any updates will                                     always take the same tlk strrefs.
                              Must be used with --start spells:xx or with startspells in a config file.
                              Only Name and SpellDesc will be sent to the spells start
                              Other spells.2da values will be sent to the default start (-s or start in config                               file)
                              A value of -1 keeps this value turned off.
                              A value <= 839 will turn it on for the whole 2da.
                              A value >839 will ignore any rows (padding) between 839 and the row ID (RID)
                              Config key: reservspells [default: -1]
  --start CSV                 Use if you want to set where individual 2das start, comma seperated.
                              Example: spells:50,classes:1000
                              Config key: start<2daname> without the <>
```

Example:

nwn_2da_tlkify -o test.tlk -d ./2da/ -a ./o2da/ -j ./json/

Will create a tlk file by the name of test in the working directory, source the 2das from 2da folder, source the json from the json folder, and output the new 2das to the o2da folder.

2) Use a config file

Sample config file, default name 2tconfig.ini
```
[General]
tlk="./Tlk input/original.tlk" # input tlk file
outtlk="test.tlk" # new tlk file
twoda="./Input 2das/" # directory for the original/input 2das
tlkjson="./Input json/" # directory for the description jsons, should be named after 2da name.json
outtwoda="./Output 2das/" # directory for output twodas
start=50 # where the tlk row additions start
```

## Json format

The name of the 2da should match the name of the json.

Each json node needs an "id" element which matches the row number within the 2da

All other elements names should match the column names of the 2da, this is case sensitive.

Example json:

```
[
	{
		"id": 1,
		"Description": "domain 1"
	},
	{
		"id": 4,
		"Description": "domain 4"
	},
	{
		"id": 6,
		"Description": "domain 6",
		"Name": "domain name"
	}
]
```

The above will change the "Description" column of a 2da for 2da rows 1, 4, and 6 and change the Name column of row 6. If the value isn't already within the tlk it will be added.

## Notes:

A starting strref of highest will append everything to the end of the tlk.

This program doesn't currently care if a strref has a value or not and will overwrite the current value.

Feel free to supply the same value multiple times, the program will only put one instance into the tlk file.

Automated options:
  --auto-plural When on fills out the Plural Columns within classes and racialtypes.2da from the Name field, the function that determines how to plural is rather basic
  --auto-adjective When on fills out ConverName field in racialtpes.2da from Name field. Basically if it ends with f, f->ven.
  --auto-iprp_spells Fills out the Name field in iprp_spells from spells.2da
  --auto-spells Fills out Name/SpellDesc from feats.2da
  --auto-iprp_feats Fills out Name from feats.2da

Automated options will function if enabled and if a value is NOT explicitly defined:

```
{
  "Name":"Orc"
},
{
  "Name":"Drow",
  "NamePlural":"Drow"
}
```

The above json will generate a Name value of Orc and NamePlural will automatically be filled with Orcs. Another Name field will be filled with "Drow" and the NamePlural field will also be "Drow"; if left to be auto filled it would had been "Drows"


## Known issues:

SpellDesc and Name field from spells.2da isn't fully supported. Depending on how you use this tool, the tlk strrefs may change from one use to the next which will cause scroll items to have the wrong description.

This can be avoided by using the --reserv-spells option.

# nwn_scrollmake

A tool to generate uti files for scrolls

## Command line

```
  -i 2DADIR                   2da DIR. This directory should contain spells.2da, classes.2da, iprp_spells.2da and des_crft_scroll.2da.
                              Config key: indir
  -j SPELL                    The path of the file spells.json. Config Key: spellsjson
  -o OUTDIR                   Output directory. Config key: outdir
  -m                          Manages des_crft_scroll.2da Will add the necessary columns/rows and fill it out.
                              Config Key: managecrft true/false
  -t TLK                      Use tlk values instead of tlk reference. Should contain the path to a tlk file. Config Key: tlkpath
  -c CONFIG                   Path to config file. [default: scrollmake.ini]
  -p PRE                      The prefix if there's no current scroll resref. Config key: prefix [default: tc_scr_]
  -d PLT                      The palette id. Config key: palID [default: 26]
```

## Config file

You can create a config file with the above Config keys, the section should be titled [Scrollmake]

## How it works

Reads classes.2da to determine what classes are able to cast spells.

Reads des_crft_scroll.2da to get the current scroll resref, if none is found it will create a scroll with the prefix defined with -p flag followed by the spell ID.

Reads spells.2da to determine what spells are castable (have a valid label, name, desc, innate level, and a casting class)

Reads through iprp_spells.2da and grabs the first item property that corresponds with the spell.

Then if it's not a raise dead, neutralize posion, remove disease, etc... the scrolls that normally don't have use restrictions it will apply the class restrictions.

For the scroll's name/description fields it uses the tlk file's strref. Optionally you can specify a tlk file (with -t) to use the tlk tables values instead.

With the -m flag, a des_crft_scroll.2da will be generated filled out with additional columns/rows and the values set to the scroll resref if the class can cast the spell. Note: If a class loses a spell, the resref won't be removed from des_crft_scroll.2da but this hardly matters because if they can't cast the spell they can't craft it.

## Advanced

### How to handle multiple item properties corresponding to the same spell:

Within a json file, which can be the same spells.2da used for nwn_2da_tlkify. You can specifiy a "scroll" field with a value of the iprp_spells row

### How to handle domain spells  

Within the above json file, you can also set "domain":true if you want the scroll to be cleric restricted even if they can't usually cast it (Example: Barkskin)

Changelog.md
* New in 0.1.3
  nwn_scrollmake: see Readme

* New in 0.1.2
  New automated options:
    --auto-plural When on fills out the Plural Columns within classes and racialtypes.2da from the Name field, the function that determines how to plural is rather basic
    --auto-adjective When on fills out ConverName field in racialtpes.2da from Name field. Basically if it ends with f, f->ven.
    --auto-iprp_spells Fills out the Name field in iprp_spells from spells.2da
    --auto-spells Fills out Name/SpellDesc from feats.2da
    --auto-iprp_feats Fills out Name from feats.2da

  --start Ability to define where each 2da's string ref starts within a tlk table.

  --reserv-spells When enabled will ensure that spells.2da Name/SpellDesc columns are always placed within the same StrRef

* New in 0.1.1

  ConverName is automatically converted to lower case for ConverNameLower in racialtypes.2da
  Name is automatically converted to lower case for Lower in classes.2da
  These options can be turned off with the new --auto-lower paramater. If a value is provided in the json for the lower case column it will continue to be used instead.
  File streams are now closed

src/private/lang.nim

proc plural*(s: string): string =
    let last = s[^1]
    let twolast = s[^2..^1]
    #if twolast == "us":
    #  result = s[0..^3]&"i"
    if twolast == "is":
      result = s[0..^3]&"es"
    elif last == 's' or last == 'x' or last == 'z' or twolast == "ch" or twolast == "sh":
      result = s & "es"
    elif last == 'f':
      result = s[0..^2]&"ves"
    elif twolast == "fe":
      result = s[0..^3]&"ves"
    elif last == 'y':
      let tolast = s[^2]
      if tolast == 'a' or tolast == 'e' or tolast == 'i' or tolast == 'o' or tolast == 'u' or tolast == 'y':
        result = s&"s"
      else:  
        result = s[0..^2]&"ies"
    #elif twolast == "on":
      #result = s[0..^3]&"a"
    else:
      result = s&"s"

proc toadjective*(s: string): string =
    let last = s[^1]
    result = s
    if last == 'f':
      result = s[0..^2]&"ven"

src/nwn_scrollmake.nim
import neverwinter/twoda, neverwinter/tlk, neverwinter/gff, docopt, options, streams, json, os, strutils, tables, parsecfg

let tlkstart = 16777216.StrRef
let doc = """
Create scrolls

Usage:
  nwn_scrollmake [options]

Options:
  -i 2DADIR                   2da DIR. This directory should contain spells.2da, classes.2da, iprp_spells.2da and des_crft_scroll.2da.
                              Config key: indir
  -j SPELL                    The path of the file spells.json. Config Key: spellsjson
  -o OUTDIR                   Output directory. Config key: outdir
  -m                          Manages des_crft_scroll.2da Will add the necessary columns/rows and fill it out.
                              Config Key: managecrft true/false
  -t TLK                      Use tlk values instead of tlk reference. Should contain the path to a tlk file. Config Key: tlkpath
  -c CONFIG                   Path to config file. [default: scrollmake.ini]
  -p PRE                      The prefix if there's no current scroll resref. Config key: prefix [default: tc_scr_]
  -d PLT                      The palette id. Config key: palID [default: 26]

 """


let args = docopt(doc)

let configf = $args["-c"]

var
 dir = $args["-i"]
 outdir = $args["-o"]
 tlkf = $args["-t"]
 mancr = args["-m"].to_bool()
 spjsonf = $args["-j"]
 dict: Config
 prefix =  $args["-p"]
 pid = parseInt($args["-d"])

if fileExists(configf):
  dict = loadConfig(configf)
  if dir == "nil":
    dir = dict.getSectionValue("Scrollmake","indir", "nil")
  if outdir == "nil":
    outdir = dict.getSectionValue("Scrollmake","outdir", "nil")
  if spjsonf == "nil":
    spjsonf = dict.getSectionValue("Scrollmake","spellsjson", "nil")
  if not mancr:
    let val = dict.getSectionValue("Scrollmake","managecrft", "false")
    if val == "true":
      mancr = true
  if tlkf == "nil":
    tlkf = dict.getSectionValue("Scrollmake","tlkpath", "nil")
  if prefix == "tc_scr_":
    prefix = dict.getSectionValue("Scrollmake","prefix", "tc_scr_")
  if pid == 26:
    pid = parseInt(dict.getSectionValue("Scrollmake","palID", "26"))

proc newProperty(self: var GffList, entry: int) =
  self.add(newGffStruct(0))
  self[entry]["ChanceToAppear", GffByte] = 100

proc classRest(self: var GffList, class: GffWord) =
  self.newProperty(self.len)
  let l = self.len-1
  self[l]["CostTable", GffByte] = 0
  self[l]["CostValue", GffWord] = 0
  self[l]["Param1", GffByte] = 255
  self[l]["Param1Value", GffByte] = 0
  self[l]["PropertyName", GffWord] = 63
  self[l]["Subtype", GffWord] = class

proc scrollTemplate(): GffRoot =

  # "create result template"
  result = newGffRoot("UTI ")

  result["AddCost", GffDWord] = 0
  result["BaseItem", GffInt] = 75
  result["Charges", GffByte] = 0
  result["Comment", GffCExoString] = ""
  result["Cost", GffDWord] = 0
  result["DescIdentified", GffCexoLocString] = newCExoLocString()
  result["Description", GffCexoLocString] = newCExoLocString()
  result["Identified", GffByte] = 1
  result["LocalizedName", GffCexoLocString] = newCExoLocString()
  result["ModelPart1", GffByte] = 0
  result["PaletteID", GffByte] = pid.GffByte
  result["Plot", GffByte] = 0
  result["Plot", GffByte] = 0
  result["Stolen", GffByte] = 0
  result["StackSize", GffWord] = 1
  result["PropertiesList", GffList] = newGffList()

  #no idea why i need to put this into a var, but it makes the compiler happy
  var props = result["PropertiesList", GffList]
  props.newProperty(0)
  # especially when i just... reassigned it here no problem
  result["PropertiesList", GffList] = props
  result["PropertiesList", GffList][0]["CostTable", GffByte] = 3
  result["PropertiesList", GffList][0]["CostValue", GffWord] = 1
  result["PropertiesList", GffList][0]["Param1", GffByte] = 255
  result["PropertiesList", GffList][0]["Param1Value", GffByte] = 0
  result["PropertiesList", GffList][0]["PropertyName", GffWord] = 15

proc isValid(cell: Cell): bool =
  if cell.isSome:
    result = true

if outdir == "nil":
  quit("No output directory specified.")

let
  classf = dir&"classes.2da"
  crftscf = dir&"des_crft_scroll.2da"
  spellf = dir&"spells.2da"
  ispellf = dir&"iprp_spells.2da"

if not fileExists(classf):
  quit("classes.2da missing")
elif not fileExists(crftscf):
  quit ("des_crft_scroll.2da missing")
elif not fileExists(spellf):
  quit ("spells.2da missing")
elif not fileExists(ispellf):
  quit ("iprp_spells.2da missing")
elif tlkf != "nil" and not fileExists(tlkf):
  quit("Tlk file missing. Note: This is optional, if you do not want to use a tlk table don't specify a tlk file.")
elif spjsonf != "nil" and not fileExists(spjsonf):
  quit("Spells.json file missing. Note: This is optional, if you do not want to use a spells.json do not specify a spells.json file.")


var
  tlkr: SingleTlk

if tlkf != "nil":
  tlkr = openFileStream(tlkf).readSingleTlk()


# Stage one: load classes
let
  classs = openFileStream(classf)
  class = classs.readTwoDA()
  classc = class.columns.find("SpellTableColumn")
classs.close

 # get class spell tables
var classes = initTable[string, seq[int]]()

for i in 0..class.high:
  let c = class[i].get()[classc]
  if isValid(c):
    discard classes.hasKeyOrPut(c.get(), @[])
    classes[c.get()].add(i)


var scrollc: seq[int]



let crftscrs = openFileStream(crftscf)
let crftscr = crftscrs.readTwoDA()
crftscrs.close


# add classes to des_crft_scroll
for i in classes.keys:
  let col =crftscr.columns.find(i)
  if col > -1:
    scrollc.add(col)
  elif mancr:
    var seq = crftscr.columns
    seq.add(i)
    crftscr.columns = seq
    for n in 0..crftscr.high:
      var row = crftscr[n].get()
      row.add(crftscr.default)
      crftscr[n] = row


proc getResRef(id: int): string =
  let row = crftscr[id]
  let rowg = row.get()
  for i in scrollc:
    let c = rowg[i]
    if isValid(c) and c.get() != "*****":
      result = toLowerAscii(c.get())
      break


let
  spells = openFileStream(spellf)
  spell = spells.readTwoDA()
spells.close

let
  ips = openFileStream(ispellf)
  ip = ips.readTwoDA()
ips.close




let
 labelc = spell.columns.find("Label")
 namec = spell.columns.find("Name")
 innatec = spell.columns.find("Innate")
 descc = spell.columns.find("SpellDesc")
 ipi = ip.columns.find("SpellIndex")


var
  domaint = initTable[int, bool]()
  ipt = initTable[int, int]()
# load in if domain spell and custom item property from row
if fileExists(spjsonf):

  let json = parseFile(spjsonf)
  for obj in items(json):
    let s = obj.hasKey("scroll")
    let d = obj.hasKey("domain")
    if s or d:
      let id = obj["id"].getInt()
      if s:
        ipt[id] = obj["scroll"].getInt()
      if d:
        domaint[id] = obj["domain"].getBool()

for i in 0..spell.high:
  let
    row = spell[i].get()
    label = row[labelc]
    name = row[namec]
    innate = row[innatec]
    desc = row[descc]
  var
    doonce = false
    scroll: GffRoot
    resref: string
    crrow: Row

  if isValid(label) and isValid(name) and isValid(innate) and isValid(desc):
    for k in classes.keys:
      let c = spell.columns.find(k)
      if c > -1 and isValid(row[c]):
        #get the ip, name, and desc
        if not doonce:
          var ipn = -1
          if ipt.hasKey(i):
            ipn = ipt[i]
          else:
            #get the first ip
            for p in 0..ip.high:
              let ipc = ip[p].get()[ipi]
              if isValid(ipc) and ipc.get() == $i:
                ipn = p
                break
            if ipn == -1:
              echo "No item property for spell: " & $i
              break
          # get the resref, also if managing des_crft_scroll create the whole row or modify it
          if crftscr[i].isSome:
            if mancr:
              crrow = crftscr[i].get()
            resref = getResRef(i)
            if resref == "":
              resref = prefix & $i
          else:
           resref = prefix & $i
           if mancr:
             crrow.add(label)
             for x in 1..<crftscr.columns.len:
               crrow.add(crftscr.default)

          doonce = true
          # actually generate the scroll here
          scroll = scrollTemplate()
          let descval = parseInt(desc.get()).StrRef
          let nameval = parseInt(name.get()).StrRef
          if tlkf != "nil" and descval >= tlkstart:
            scroll["DescIdentified", GffCexoLocString].entries[0] = $tlkr[descval-tlkstart].get()
          else:
            scroll["DescIdentified", GffCexoLocString].strRef = descval
          if tlkf != "nil" and nameval >= tlkstart:
            scroll["LocalizedName", GffCexoLocString].entries[0] = $tlkr[nameval-tlkstart].get()
          else:
            scroll["LocalizedName", GffCexoLocString].strRef = nameval

          scroll["PropertiesList", GffList][0]["Subtype", GffWord] = ipn.GffWord

          scroll["Tag", GffCExoString] = toUpperAscii(resref)
          scroll["TemplateResRef", GffResRef] = resref.GffResRef

          if domaint.hasKey(i) and domaint[i]:
            var cleric = scroll["PropertiesList", GffList]
            cleric.classRest(2)
            scroll["PropertiesList", GffList] = cleric
        # end get the ip, name, and desc
        #add classe use restrictions here, except on scrolls that don't accept use restrictions
        if not (i == 70 or i == 97 or i == 152 or i == 142 or i == 153 or  i == 147 or i == 126):
          var props = scroll["PropertiesList", GffList]
          for class in classes[k]:
            props.classRest(class.GffWord)
          scroll["PropertiesList", GffList] = props
        if mancr:
          crrow[crftscr.columns.find(k)] = some(resref)
    #generated a scroll lets add the necessary row(s) to the 2da
    if doonce and mancr:
      if i > crftscr.high:
        for h in crftscr.high+1..i:
          var newRow: Row
          # we have to assign the columns too
          for x in 0..<crftscr.columns.len:
            newRow.add(crftscr.default)
          crftscr[h] = newRow
      #now we add the row we created for our scroll
      crftscr[i] = crrow



    # output
    if not isNil(scroll):
      let item = openFileStream(outdir&resref&".uti", fmWrite)
      item.write(scroll)
      item.close

if mancr:
  let crftos = openFileStream(outdir&"des_crft_scroll.2da", fmWrite)
  crftos.writeTwoDA(crftscr)

src/nwn_2da_tlkify.nim
import neverwinter/twoda, neverwinter/tlk, docopt, options, streams, json, parsecfg, os, parseutils, strutils, private/lang  #streams, options, json, parsecfg, re, os, strutils, typetraits

let doc = """
Update 2da and tlk files for NWN 1

Usage:
  nwn_2da_tlkify [options]

Options:
  -i TLK                      Input TLK file, Config key: tlk
  -s START                    Starting StrRef, Config key: start [default: highest]
  -o OUTTLK                   Output TLK file file, Config key: outtlk
  -d INDIR                    2da input directory, Config key: twoda
  -a OUTDIR                   2da output directory Config key: outtwoda
  -j JSONDIR                  Json Input directory, Config key: tlkjson
  -c CONFIG                   Configuration file [default: 2tconfig.ini]
  --auto-lower SEL            Turns on, off, or on for a specific 2da the automatically setting
                              the Lower field in classes or ConverNameLower field in racialtypes
                              Possible selections: all, off, class, and race.
                              Config key: autolower [default: all]
  --auto-plural PLU           Turns on, off, or on for a specific 2da the automatically setting
                              the Plural field in classes or NamePlural field in racialtypes
                              Possible selections: all, off, class, and race.
                              Config key: autoplural [default: all]
  --auto-adjective            Turns on auto setting of the ConverName from Name in racialtypes.
                              Config key: autoadjective true/false
  --auto-iprp_spells ISP      Sets the iprp_spells row to start autofill from the spells.2da.
                              Config key: autoiprp_spells [default: -1]
  --auto-spells SID           Sets the spells row to start autofill from feat.2da for Name and SpellDesc columns.
                              Config key: autospells [default: -1]
  --auto-iprp_feats FID       Sets the iprp_feats row to start autofill from feat.2da for the Name column.
                              Config key: autoiprp_feats [default: -1]
  --reserv-spells RID         Reserves space for spells.2da Name/SpellDesc columns so that any updates will always
                              take the same tlk strrefs.
                              Must be used with --start spells:xx or with startspells in a config file.
                              Only Name and SpellDesc will be sent to the spells start
                              Other spells.2da values will be sent to the default start (-s or start in config file)
                              A value of -1 keeps this value turned off.
                              A value <= 839 will turn it on for the whole 2da.
                              A value >839 will ignore any rows (padding) between 839 and the row ID (RID)
                              Config key: reservspells [default: -1]
  --start CSV                 Use if you want to set where individual 2das start, comma seperated.
                              Example: spells:50,classes:1000
                              Config key: start<2daname> without the <>
"""

let args = docopt(doc)


proc find(self: SingleTlk, str: string): int =
  result = -1
  for i in 0..self.highest():
    let entry = self[i.StrRef]
    if entry.isSome and entry.get().hasValue and $entry.get() == str:
      result = i
      break


var
 inputfile = $args["-i"]
 dict: Config
 outtlk = $args["-o"]
 injson = $args["-j"]
 intwoda = $args["-d"]
 outtwoda = $args["-a"]
 starts = $args["-s"]
 autol = $args["--auto-lower"]
 autop = $args["--auto-plural"]
 autoa = args["--auto-adjective"].to_Bool
 autoip = parseInt($args["--auto-iprp_spells"])
 autosp = parseInt($args["--auto-spells"])
 autoft = parseInt($args["--auto-iprp_feats"])
 rsvspl = parseInt($args["--reserv-spells"])
 start = initTable[string, int]()



let
  confl = $args["-c"]
  configExist = fileExists(confl)


if configExist:
  dict = loadConfig(confl)
  if outtlk == "nil":
    outtlk = dict.getSectionValue("General","outtlk", "nil")
  if injson == "nil":
    injson = dict.getSectionValue("General","tlkjson", "nil")
  if intwoda == "nil":
    intwoda = dict.getSectionValue("General","twoda", "nil")
  if outtwoda == "nil":
    outtwoda = dict.getSectionValue("General","outtwoda", "nil")
  if inputfile == "nil":
    inputfile = dict.getSectionValue("General","tlk", "nil")
  if starts == "highest":
    starts = dict.getSectionValue("General","start", "highest")
  if autol == "all":
    autol =  dict.getSectionValue("General","autolower", "all")
  if autop == "all":
    autop =  dict.getSectionValue("General","autoplural", "all")
  if not autoa:
    let auto = dict.getSectionValue("General","autoadjective", "false")
    if auto == "true":
        autoa = true
  if autoip == -1:
    autoip = parseInt(dict.getSectionValue("General","autoiprp_spells", "-1"))
  if autosp == -1:
    autosp = parseInt(dict.getSectionValue("General","autospells", "-1"))
  if autoft == -1:
    autoft = parseInt(dict.getSectionValue("General","autoiprp_feats", "-1"))
  if rsvspl == -1:
    rsvspl = parseInt(dict.getSectionValue("General","reservspells", "-1"))

start["default"] = -1 #initialize
if outtlk == "nil":
  quit("Error: No output defined for tlk.")
if injson == "nil":
  quit("Error: No input directory defined for json.")
if intwoda == "nil":
  quit("Error: No input directory defined for 2das.")
if outtwoda == "nil":
  quit("Error: No output directory defined for 2das.")

if (parseInt(starts, start["default"]) == 0 and starts != "highest"):
  quit("Error: tlk starting strref isn't a valid digit or highest")

var state: SingleTlk
if inputfile != "nil":
  if fileExists(inputfile):
    state  = openFileStream(inputfile).readSingleTlk()
  else:
    quit("Error: Tlk input file supplied but does not exist.")
else:
  starts = "" #initilize start here for new tlks
  state = newSingleTlk()

if starts == "highest":
  start["default"] = state.highest()+1

doAssert(start["default"] >= 0, "Error: Startting strref should not be negative.")

const tlkoffset = 16777216
var
  row: Row
  tblstream: Table[string, TwoDA]
  td_ipsp: seq[int]
  td_spellsn: seq[int]
  td_spellsd: seq[int]
  td_ipft: seq[int]


for i in split($args["--start"], ','):
  if i == "nil":
    break
  let csvi = split(i, ":")
  start[csvi[0]] = parseInt(csvi[1])


proc UpdateRAndTSpec(self: Row, colID: int, value: string, tlksr: int) =
  row[colID] = some($(tlksr+tlkoffset))
  state[tlksr.StrRef] = value

proc UpdateRAndT(self: Row, colID: int, value: string, startp: string) =
  let strref = state.find(value)
  if strref > -1:
    row[colID] = some($(strref+tlkoffset))
  else:
    UpdateRAndTSpec(row, colID, value, start[startp])
    start[startp] += 1

proc SafeAddToRow(self: TwoDA, row: Row, col: string, val: string, startp: string) =
   let colID = self.columns.find(col)
   if colID == -1:
     echo "Column name "&col&" not found"
   else:
     row.UpdateRAndT(colID, val, startp)

proc ResrvSpells(self: Row, colID: int, value: string, rowID: int, offset: int) =
  if rsvspl > -1:
    if rsvspl <= 839 or rowID <= 839:
      row.UpdateRAndTSpec(colID, value, start["spells"]+rowID*2+offset)
    else:
      row.UpdateRAndTSpec(colID, value, start["spells"]+(840+rowID-rsvspl)*2+offset)


for file in walkFiles(injson&"*.json"):
  let filesplit = splitFile(file)
  if not fileExists(intwoda&filesplit.name&".2da"):
    echo "Json found for " & filesplit.name & " but no corresponding 2da."
  else:
    let twodas = openFileStream(intwoda&filesplit.name&".2da")
    let twoda = twodas.readTwoDA()
    twodas.close
    let json = parseFile(file)
    var startp: string
    if start.hasKey(filesplit.name):
      startp = filesplit.name
    else:
       if configExist:
        let startcs = dict.getSectionValue("General","start"&filesplit.name, "-1")
        if startcs != "-1":
          startp = filesplit.name
          start[filesplit.name] = parseInt(startcs)

       if not start.hasKey(filesplit.name):
         startp = "default"
    var
      spname: int
      spdesc: int
    if rsvspl > -1 and filesplit.name == "spells":
      if start.hasKey("spells"):
        spname = twoda.columns.find("Name")
        spdesc = twoda.columns.find("SpellDesc")
        startp = "default" #redirect all non-name/spelldesc to default
      else:
        rsvspl = -1
    for itm in items(json):

      if itm.hasKey("id"):
        let rowID = itm["id"].getInt()
        row = twoda[rowID].get()
        if filesplit.name == "iprp_spells" and autoip != -1 and rowID >= autoip and itm.hasKey("Name"):
          td_ipsp.add(rowID)
        elif filesplit.name == "spells":
          if itm.hasKey("Name"):
            if autosp != -1 and rowID >= autosp:
              td_spellsn.add(rowID)
            row.ResrvSpells(spname, itm["Name"].getStr(), rowID, 0)
          if itm.hasKey("SpellDesc"):
            row.ResrvSpells(spdesc, itm["SpellDesc"].getStr(), rowID, 1)
            if autosp != -1 and rowID >= autosp:
              td_spellsd.add(rowID)
        elif filesplit.name == "iprp_feats" and autoft != -1 and rowID >= autoft and itm.hasKey("Name"):
          td_ipft.add(rowID)

        for p in pairs(itm):
          if $p.key == "id" or (rsvspl > -1 and filesplit.name == "spells" and ($p.key == "Name" or $p.key == "SpellDesc")):
            continue
          var colID = twoda.columns.find($p.key)
          if colID == -1:
            echo "Column name " & $p.key & " not found in " & filesplit.name
            continue


          row.UpdateRAndT(colID, p.val.getStr(), startp)


        if filesplit.name == "racialtypes":
          if (autol == "all" or autol == "race") and not itm.hasKey("ConverNameLower"):
            if itm.hasKey("ConverName"):
              SafeAddToRow(twoda, row, "ConverNameLower", itm["ConverName"].getStr().toLowerAscii, startp)
            elif autoa and itm.hasKey("Name"):
              SafeAddToRow(twoda, row, "ConverNameLower", itm["Name"].getStr().toadjective().toLowerAscii, startp)
          if (autop == "all" or autop == "race") and itm.hasKey("Name") and not itm.hasKey("NamePlural"):
            SafeAddToRow(twoda, row, "NamePlural", itm["Name"].getStr().plural(), startp)
          if autoa and itm.hasKey("Name") and not itm.hasKey("ConverName"):
            SafeAddToRow(twoda, row, "ConverName", itm["Name"].getStr().toadjective(), startp)
        elif filesplit.name == "classes":
          if itm.hasKey("Name"):
            if (autol == "all" or autol == "class") and not itm.hasKey("Lower"):
              SafeAddToRow(twoda, row, "Lower", itm["Name"].getStr().toLowerAscii, startp)
            if (autop == "all" or autop == "class") and not itm.hasKey("Plural"):
              SafeAddToRow(twoda, row, "Plural", itm["Name"].getStr().plural(), startp)

        twoda[rowID]=row

      else:
        echo "This does not have a valid id."

    tblstream[filesplit.name] = twoda

proc getTwoDA(name: string): TwoDA =
  if tblstream.hasKey(name):
    result = tblstream[name]
  elif fileExists(intwoda&name&".2da"):
    let tmps = openFileStream(intwoda&name&".2da")
    result = tmps.readTwoDA()
    tmps.close


if tblstream.hasKey("feat"):
  if autoft > -1:
    var updtwoda = getTwoDA("iprp_feats")
    let
      colID = updtwoda.columns.find("Name")
      colFeat = updtwoda.columns.find("FeatIndex")
      ftName = tblstream["feat"].columns.find("FEAT")

    if not isNil(updtwoda):
      for i in autoft..updtwoda.high:
        if not td_ipft.contains(i):
          row = updtwoda[i].get()
          if row[colFeat].isSome:
            let featidx = parseInt(row[colFeat].get())
            if tblstream["feat"][featidx].isSome:
                row[colID] = tblstream["feat"][featidx].get()[ftName]
                updtwoda[i] = row

      tblstream["iprp_feats"] = updtwoda

  if autosp > -1:
    var updtwoda = getTwoDA("spells")

    let
      colID = updtwoda.columns.find("Name")
      colSpell = updtwoda.columns.find("SpellDesc")
      colFeat = updtwoda.columns.find("FeatID")
      ftName = tblstream["feat"].columns.find("FEAT")
      ftDesc = tblstream["feat"].columns.find("DESCRIPTION")
    if not isNil(updtwoda):
      for i in autosp..updtwoda.high:
        if not td_spellsn.contains(i) or not td_spellsd.contains(i):
          row = updtwoda[i].get()
          if row[colFeat].isSome:
            let featidx = parseInt(row[colFeat].get())
            if tblstream["feat"][featidx].isSome:
              if not td_spellsn.contains(i):
                row[colID] = tblstream["feat"][featidx].get()[ftName]
              if not td_spellsd.contains(i):
                row[colSpell] = tblstream["feat"][featidx].get()[ftDesc]
          updtwoda[i] = row

      tblstream["spells"] = updtwoda

if autoip > -1 and tblstream.hasKey("spells"):
  var updtwoda = getTwoDA("iprp_spells")

  let colID = updtwoda.columns.find("Name")
  let colSpell = updtwoda.columns.find("SpellIndex")
  let spName = tblstream["spells"].columns.find("Name")
  if not isNil(updtwoda):
    for i in autoip..updtwoda.high:
      if not td_ipsp.contains(i):
        row = updtwoda[i].get()
        if row[colSpell].isSome:
          let spellidx = parseInt(row[colSpell].get())
          if tblstream["spells"][spellidx].isSome:
            row[colID] = tblstream["spells"][spellidx].get()[spName]
            updtwoda[i] = row

    tblstream["iprp_spells"] = updtwoda

for k, v in tblstream.pairs:
  let outda = newFileStream(outtwoda&k&".2da", fmWrite)
  outda.writeTwoDA(v)
  outda.close

let output = newFileStream(outtlk, fmWrite)

output.writeTlk(state)
